### PR TITLE
feat: Improve managed node group bootstrap revisited

### PR DESCRIPTION
**Pull request by stevehipwell**
# PR o'clock

## Description

This PR could replace https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1433.

The managed node group bootstrap pattern has been modified to set the `/etc/eks/bootstap.sh` environment variables to control the bootstrap process. To do this a new `bootstrap_env` variable has been added which sets env variables to be sourced into `/etc/eks/bootstrap.sh` from a new `/etc/profile.d/eks-bootstrap.sh` file. For power users `pre_userdata` can append additional values or dynamic functionality to this file.

### Checklist

- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by ArchiFleKs** [comment]
How does this work with custom AMI without merging userdata and calling boostrap manually, are this labels still set on the launch template user data:

```
/etc/eks/bootstrap.sh mycluster --kubelet-extra-args '--node-labels=eks.amazonaws.com/nodegroup-image=ami-072fbef2a16895c26,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/sourceLaunchTemplateVersion=1,eks.amazonaws.com/nodegroup=my-cluster-b-20210713121718836600000009,role=worker,eks.amazonaws.com/sourceLaunchTemplateId=lt-0034c7256eaa78b73' --b64-cluster-ca $B64_CLUSTER_CA --apiserver-endpoint $API_SERVER_URL --dns-cluster-ip $K8S_CLUSTER_DNS_IP
```

Which are the automatically merged bootstrap script when not using custom ami

**Comment by stevehipwell** [comment]
@ArchiFleKs this doesn't change the current behaviour, so I'm not quite sure what the question is? If an AMI is provided then we call `/etc/eks/bootstrap.sh mycluster` at the end of our user data, if it's not our user data is merged with the default userdata which makes this call.

**Comment by stevehipwell** [comment]
@ArchiFleKs I think I see your point, let me make a change.

**Comment by stevehipwell** [comment]
@ArchiFleKs this looks like a defect in the current release for custom AMIs? I can see there is also a defect in my PR as `bootstrap_env.KUBELET_EXTRA_ARGS` will loose out to the `--kubelet-extra-args` input so I'll fix this.

**Comment by stevehipwell** [comment]
@ArchiFleKs based on the code and AWS docs I suspect you would need to replicate the bootstrap args or set the env variables for a custom AMI to work as expected.

**Comment by stevehipwell** [comment]
@ArchiFleKs could you take a look at the changes to fix the behaviour when no `ami_id` is passed in? I think that a separate PR should be opened to set the default bootstrap arguments when `ami_id` is set.

**Comment by stevehipwell** [comment]
@daroga0002 is this they type of generic you were looking for? The only pain point with a generic solution like this is that the default values will be overridden as a single value due to TF limitations.

**Comment by ArchiFleKs** [comment]
> @ArchiFleKs could you take a look at the changes to fix the behaviour when no `ami_id` is passed in? I think that a separate PR should be opened to set the default bootstrap arguments when `ami_id` is set. 

I think with the if statement It should be ok with standard AMI

**Comment by ArchiFleKs** [comment]
How would someone wanting to use `containerd` and max pod with VPC CNI > 1.9  would do it with this PR compare to the older one ?

**Comment by stevehipwell** [comment]
> How would someone wanting to use containerd and max pod with VPC CNI > 1.9 would do it with this PR compare to the older one ?

@ArchiFleKs the following config would do this assuming that you wanted 110 pods per node.

```terraform
node_groups = {
  my_group = {
    bootstrap_env = {
      CONTAINER_RUNTIME = "containerd"
      USE_MAX_PODS      = false
    }
    kubelet_extra_args = "--max-pods=110"
  }
}
```

**Comment by daroga0002** [comment]
@stevehipwell  just to let you know I seeing your good job :tada: and have your PRs on list to review (and test).


I will try to find a time soon for this

**Comment by stevehipwell** [comment]
@daroga0002 would you like me to close https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1433 so as to not confuse matters? I think this one is preferred by the maintainers?

**Comment by daroga0002** [comment]
> @daroga0002 would you like me to close #1433 so as to not confuse matters? I think this one is preferred by the maintainers?

lets keep this open as I will took look on both

**Comment by ArchiFleKs** [comment]
> > How would someone wanting to use containerd and max pod with VPC CNI > 1.9 would do it with this PR compare to the older one ?
> 
> 
> 
> @ArchiFleKs the following config would do this assuming that you wanted 110 pods per node.
> 
> 
> 
> ```terraform
> 
> node_groups = {
> 
>   my_group = {
> 
>     bootstrap_env = {
> 
>       CONTAINER_RUNTIME = "containerd"
> 
>       USE_MAX_PODS      = false
> 
>     }
> 
>     kubelet_extra_args = "--max-pods=110"
> 
>   }
> 
> }
> 
> ```

Just tested this, working as expected, I think this PR is better than the old one because this allow to use pretty much any environments variable available in eks bootstrap script without having to do another PR here. 

True it is more obvious to user but I think we can give example instead.

**Comment by jaimehrubiks** [comment]
I like it. Just need to make sure that it satisfies the needs for people using custom AMI, but for the rest, it looks good

**Comment by stevehipwell** [comment]
@jaimehrubiks I'd take #1580 first and then add these changes on top of that; so custom AMIs are supported correctly.

**Comment by FabioAntunes** [comment]
Hey folks quick question, without this PR is there any other way to enable `containerd`  in our managed nodes?

**Comment by stevehipwell** [comment]
@FabioAntunes you can implement the code in this PR directly into your userdata to achieve the same behaviour, this is what I'm doing until the PR is merged.

**Review by daroga0002 — APPROVED** [review]
@stevehipwell  thank you for contribution 🎉 

I have reviewed and tested this and works fine. 

Diff between new and old approach in userdata:
```
13c13,15
< export ADDITIONAL_KUBELET_EXTRA_ARGS=""
---
> export CONTAINER_RUNTIME="containerd"
> export USE_MAX_PODS="false"
> export ADDITIONAL_KUBELET_EXTRA_ARGS="--max-pods=110"
32c34
```
+ tag on launch template which I omitted in diff.

**Comment by daroga0002** [comment]
@antonbabenko  lets merge this to master

**Comment by antonbabenko** [comment]
Here we go! 🎉 

[v17.23.0](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/v17.23.0) has been just released.

**Comment by stevehipwell** [comment]
Thank you @daroga0002 & @antonbabenko, I'll open a PR shortly to support Bottlerocket in MNGs. Is there any news on a refactored pattern for un-managed launch template nodes?

**Comment by daroga0002** [comment]
> Is there any news on a refactored pattern for un-managed launch template nodes?

I dont think so, in general there was quite good PR:
https://github.com/terraform-aws-modules/terraform-aws-eks/pull/858
but now it is outdated and probably require some effort to fix and port new features into that.

Additional consideration is how to make it more consistent with node_groups submodule as some code probably can be shared (like launch template).

**Comment by stevehipwell** [comment]
@daroga0002 it might make more sense to add a new input and create a sibling module to the MNG to create non-MNGs without needing a breaking change. The launch template code could also then be extracted into a shared module, although that might want to wait until both are stable.

**Comment by daroga0002** [comment]
> it might make more sense to add a new input and create a sibling module to the MNG to create non-MNGs without needing a breaking change.

maybe, but from other side it will be nightmare to maintain it.

I think we need here finally release some major breaking change and fixing a lot of `legacy` mistakes

**Comment by stevehipwell** [comment]
> maybe, but from other side it will be nightmare to maintain it.

I'd propose doing this first and then doing the big breaking changes refactor to remove legacy issues and side effects. It might even line it up with the future Terraform auto state mapping (if the removal of count isn't already supported) so it's only a breaking change if you're using the legacy patterns and side effects. I'd be keen to be a part of this work.

**Comment by jurabek** [comment]
Hi, @ArchiFleKs is it possible to set `CONTAINER_RUNTIME=containerd` somehow when you use `worker_groups`? I tried to set  `bootstrap_env` to `worker_groups` itself 
```
bootstrap_env = {
      CONTAINER_RUNTIME = "containerd"
      USE_MAX_PODS      = false
    }
```
I really appreciate any suggestion here, 
Thanks

**Comment by ArchiFleKs** [comment]
> Hi, @ArchiFleKs is it possible to set `CONTAINER_RUNTIME=containerd` somehow when you use `worker_groups` instead `node_groups`? I tried to set `bootstrap_env` to `worker_groups` itself
> 
> ```
> bootstrap_env = {
>       CONTAINER_RUNTIME = "containerd"
> }
> ```
> 
> I really appreciate any suggestion here, Thanks

Unsure I'm not using work_groups anymore, maybe @stevehipwell can give you a better answer but I don't think you can based on [this](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/templates/userdata.sh.tpl)

**Comment by stevehipwell** [comment]
@jurabek `bootstrap_env` is currently only supported in MNGs, but I am planning on opening a PR to add it to worker groups. In the meantime you can use the `pre_userdata` to set the variables directly (this is safe as only MNGs merge userdata.

```hcl
pre_userdata = <<-EOT
  export CONTAINER_RUNTIME="containerd"
EOT
```

**Comment by jurabek** [comment]
@stevehipwell that worked for me, thank you very much.

**Pull request merged**
